### PR TITLE
(api) Add option to enforce use of scopes (`all/read`, `all/write`, etc.)

### PR DESCRIPTION
**Pull request by Leobouloc**
## Purpose

The current state of Ralph allows to restrict users by authority, but does not allow 1. An admin user 2. Finer access control (read, write). This PR aims to solve this issue by implementing `scopes` (a field already present in user accounts) which can allow restricted access.

## Proposal

The scopes proposed are a slight variation on the scopes defined by the [xAPI standard](https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Communication.md#details-15):

```
    "statements/write",
    "statements/read/mine",
    "statements/read",
    "state/write",
    "state/read",
    "define",
    "profile/write",
    "profile/read",
    "all/read",
    "all",
```

NB: This PR also proposes some cleaning (renaming tests and factorizing code in tests).

TODO:
- [x] restrict endpoint access by scopes (`RALPH_LRS_RESTRICT_BY_SCOPE`)
- [x] deal with  `statements/read/mine` scope (`RALPH_LRS_RESTRICT_BY_AUTHORITY` must be used) (specific test !)
- [x] test restrictions for all endpoints (GET, POST, PUT) AND for both auth methods (OIDC, and basic auth)
- [x] check that `LRS_RESTRICT_BY_AUTHORITY==True` if using scopes, and test this behavior

### CODE REVIEW

File: src/ralph/api/auth/user.py [hunk 1]
```diff
@@ -18,6 +19,42 @@
...
+                "statements/read",
+                "state/read",
+                "state/write",
+                "define",
+                "profile/read",
+                "profile/write",
+                "all/read",
+                "all",
+            },
+        }
```
Comment (Leobouloc) [inline src/ralph/api/auth/user.py:49]: > Wouldn't it be possible to define them with a wildcard? E.g. statements/read/* and */read/*?
(initially commented by @wilbrdt)

I prefered this explicit syntax as it seemed clearer when reading the code, but I can change this. Anyone else has an opinion ?

Comment (wilbrdt) [reply src/ralph/api/auth/user.py:49]: I'm fine with this explicit syntax!

File: tests/api/auth/test_basic.py [hunk 2]
```diff
@@ -217,14 +220,17 @@ def test_get_whoami_correct_credentials(basic_auth_test_client, fs):
...
     fs.create_file(auth_file_path, contents=STORED_CREDENTIALS)
-    get_authenticated_user.cache_clear()
+    get_basic_auth_user.cache_clear()
 
     response = basic_auth_test_client.get(
         "/whoami", headers={"Authorization": f"Basic {credentials}"}
     )
 
     assert response.status_code == 200
-    assert response.json() == {
```
Comment (quitterie-lcs) [inline tests/api/auth/test_basic.py:227]: Why did you change and split this assertion ?

File: tests/api/test_statements_get.py [hunk 3]
```diff
@@ -739,6 +745,168 @@ def test_api_statements_get_invalid_query_parameters(auth_credentials, id_param)
...
+        ([], False),
+    ],
+)
+def test_api_statements_get_scopes(
+    monkeypatch, fs, es, auth_method, scopes, is_authorized
+):
+    """Test that getting statements behaves properly according to user scopes."""
+    # pylint: disable=invalid-name
+    # pylint: disable=too-many-locals
+    # pylint: disable=too-many-arguments
```
Comment (quitterie-lcs) [inline tests/api/test_statements_get.py:775]: Can't you stack them into one line ?

File: src/ralph/api/auth/user.py (hunk 1 above)
Comment (jmaupetit) [reply src/ralph/api/auth/user.py:49]: Mybe `Scope` and expanded scopes can be defined as `Enum`s listing all possible (expanded) scopes. This will be easier to manipulate if we should refer to a particular (expanded) scope. WDYT?

File: src/ralph/api/routers/statements.py [hunk 4]
```diff
@@ -137,10 +139,12 @@ def strict_query_params(request: Request):
 
 @router.get("")
 @router.get("/")
-# pylint: disable=too-many-arguments, too-many-locals
 async def get(
     request: Request,
-    current_user: Annotated[AuthenticatedUser, Depends(get_authenticated_user)],
+    current_user: Annotated[
+        AuthenticatedUser,
+        Security(get_authenticated_user, scopes=["statements/read/mine"]),
```
Comment (jmaupetit) [inline src/ralph/api/routers/statements.py:146]: Following my previous comment, this may be expressed as:

```suggestion
        Security(get_authenticated_user, scopes=[SCOPES.STATEMENTS_READ_MINE]),
```

File: src/ralph/conf.py [hunk 5]
```diff
@@ -210,5 +213,18 @@ def LOCALE_ENCODING(self) -> str:  # pylint: disable=invalid-name
...
+    @root_validator(allow_reuse=True)
+    @classmethod
+    def check_restriction_compatibility(cls, values):
+        """Raise an error if scopes are being used without authority restriction."""
+        if values.get("LRS_RESTRICT_BY_SCOPES") and not values.get(
+            "LRS_RESTRICT_BY_AUTHORITY"
+        ):
+            raise ConfigurationException(
+                "`LRS_RESTRICT_BY_AUTHORITY` must be set to `True` if using "
+                "`LRS_RESTRICT_BY_SCOPES=True`"
```
Comment (jmaupetit) [inline src/ralph/conf.py:225]: Not sure back-quotes are required for exception strings:

```suggestion
                "LRS_RESTRICT_BY_AUTHORITY must be set to True if using "
                "LRS_RESTRICT_BY_SCOPES=True"
```

File: tests/api/test_statements_get.py (hunk 3 above)
Comment (jmaupetit) [reply tests/api/test_statements_get.py:775]:
```suggestion
    # pylint: disable=invalid-name,too-many-locals,too-many-arguments
```

**Review by jmaupetit — COMMENTED** [review]
Looks great! My suggestion about using Enum for scopes can be handled in a new PR (if you are ok with this).

File: tests/api/test_statements_put.py [hunk 6]
```diff
@@ -515,3 +526,74 @@ async def test_api_statements_put_with_forwarding(
...
+        (["all/read"], False),
+        (["statements/read/mine"], False),
+        (["profile/write"], False),
+        ([], False),
+    ],
+)
+def test_api_statements_put_scopes(
+    monkeypatch, fs, es, auth_method, scopes, is_authorized
+):
+    """Test that getting statements behaves properly according to user scopes."""
```
Comment (SergioSim) [inline tests/api/test_statements_put.py:547]:
```suggestion
    """Test that writing statements behaves properly according to user scopes."""
```

**Review by SergioSim — COMMENTED** [review]
This looks great :+1: 
Have only minor questions/suggestions)

image

File: src/ralph/api/auth/user.py [hunk 7]
```diff
@@ -18,6 +19,42 @@
 ]
 
 
+class UserScopes(FrozenSet[Scope]):
+    """Scopes available to users."""
+
+    @lru_cache()
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/user.py:25]: lets go for maxsize

File: tests/api/test_statements_post.py [hunk 8]
```diff
@@ -614,3 +627,74 @@ async def test_api_statements_post_list_with_forwarding(
...
+        (["all/read"], False),
+        (["statements/read/mine"], False),
+        (["profile/write"], False),
+        ([], False),
+    ],
+)
+def test_api_statements_post_scopes(
+    monkeypatch, fs, es, auth_method, scopes, is_authorized
+):
+    """Test that getting statements behaves properly according to user scopes."""
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply tests/api/test_statements_post.py:648]: Replaced by `posting` and `putting`

File: tests/api/test_statements_put.py (hunk 6 above)
Comment (Leobouloc) [reply tests/api/test_statements_put.py:547]: replaced by `putting`

File: src/ralph/api/auth/user.py [hunk 9]
```diff
@@ -18,6 +19,42 @@
...
+                "profile/read",
+                "profile/write",
+                "all/read",
+                "all",
+            },
+        }
+
+        expanded_user_scopes = set()
+        for scope in self:
+            expanded_user_scopes.update(expanded_scopes.get(scope, {scope}))
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/user.py:53]: Yes, but I don't see the use for this as required scopes in an endpoint are always "leafs" of the tree.

File: src/ralph/api/auth/user.py [hunk 10]
```diff
@@ -18,6 +19,42 @@
 ]
 
 
+class UserScopes(FrozenSet[Scope]):
+    """Scopes available to users."""
+
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/user.py:24]: Yes. I can't remember why I chose this 🙃

File: src/ralph/api/auth/__init__.py [hunk 11]
```diff
@@ -1,12 +1,11 @@
 """Main module for Ralph's LRS API authentication."""
 
-from ralph.api.auth.basic import get_authenticated_user as get_basic_user
-from ralph.api.auth.oidc import get_authenticated_user as get_oidc_user
+from ralph.api.auth.basic import get_basic_auth_user
+from ralph.api.auth.oidc import get_oidc_user
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/__init__.py:4]: I found that using aliases was making the code and dev process more complicated than necessary, which is why I changed this. 

Since there are currently no constraints for homogeneity (no base class), I would argue we are not really breaking anything here. We could make the auth methods more homogenous but this would require a bit more work which should probably be handled in a separate PR.

File: tests/api/test_statements_put.py [hunk 12]
```diff
@@ -515,3 +526,74 @@ async def test_api_statements_put_with_forwarding(
 
     # Stop receiving LRS client
     await lrs_context.__aexit__(None, None, None)
+
+
+@responses.activate
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply tests/api/test_statements_put.py:531]: It seems so, for OIDC mocking.

File: src/ralph/api/auth/__init__.py [hunk 13]
```diff
@@ -1,12 +1,11 @@
...
 # At startup, select the authentication mode that will be used
-get_authenticated_user = (
-    get_oidc_user
-    if settings.RUNSERVER_AUTH_BACKEND == settings.AuthBackends.OIDC
-    else get_basic_user
-)
+if settings.RUNSERVER_AUTH_BACKEND == settings.AuthBackends.OIDC:
+    get_authenticated_user = get_oidc_user
+else:
+    get_authenticated_user = get_basic_auth_user
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/api/auth/__init__.py:11]: I don't have a strong opinion on this.

File: src/ralph/api/auth/user.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/user.py:25]: I think a fixed value of 1024 makes sense here as there is a limited number of combinations possible for scope input values.

File: setup.cfg [hunk 14]
```diff
@@ -142,7 +142,7 @@ match = ^(?!(setup)\.(py)$).*\.(py)$
 [isort]
 known_ralph=ralph
 sections=FUTURE,STDLIB,THIRDPARTY,RALPH,FIRSTPARTY,LOCALFOLDER
-skip_glob=venv
+skip_glob=venv,*/.conda/*
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply setup.cfg:145]: Yes I think it  should apply to all developpers

File: src/ralph/api/auth/oidc.py [hunk 15]
```diff
@@ -143,7 +145,20 @@ def get_authenticated_user(
...
+    # Restrict access by scopes
+    if settings.LRS_RESTRICT_BY_SCOPES:
+        for requested_scope in security_scopes.scopes:
+            is_auth = user.scopes.is_authorized(requested_scope)
+            if not is_auth:
+                raise HTTPException(
+                    status_code=status.HTTP_401_UNAUTHORIZED,
+                    detail=f'Access not authorized to scope: "{requested_scope}".',
+                    headers={"WWW-Authenticate": "Basic"},
+                )
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/oidc.py:162]: I agree. This factoring is done in the following PR:

https://github.com/openfun/ralph/blob/45f0231d77a74fdc29de8dbf2430b85128524070/src/ralph/api/auth/__init__.py#L41

File: src/ralph/conf.py (hunk 5 above)
Comment (Leobouloc) [reply src/ralph/conf.py:225]: Removed them !

File: tests/api/test_statements_get.py (hunk 3 above)
Comment (Leobouloc) [reply tests/api/test_statements_get.py]: done !

File: tests/api/auth/test_basic.py (hunk 2 above)
Comment (Leobouloc) [reply tests/api/auth/test_basic.py:227]: Following recent remarks on the use of sets vs lists, I decided to use sets to store scopes, which means we cannot assume an order in the scopes we retrieve.

File: src/ralph/api/auth/basic.py [hunk 16]
```diff
@@ -182,4 +183,16 @@ def get_authenticated_user(
             headers={"WWW-Authenticate": "Basic"},
         )
 
-    return AuthenticatedUser(scopes=user.scopes, agent=user.agent)
+    user = AuthenticatedUser(scopes=UserScopes(user.scopes), agent=user.agent)
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/basic.py]: Change made

File: src/ralph/api/auth/user.py (hunk 1 above)
Comment (Leobouloc) [reply src/ralph/api/auth/user.py:51]: To be changed later : https://github.com/openfun/ralph/issues/479

File: src/ralph/api/auth/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/__init__.py:4]: closed by thumbsup

File: src/ralph/api/auth/user.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/user.py:26]: change made

File: src/ralph/api/auth/user.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/user.py:53]: hmm yes I see what you mean,  hadn't thought of that.

But I feel like even in the event of scopes not being "leafs" (for example with a scope `"read/all"`), it would be preferable not to infer that `"statements/read + state/read +  profile/read" <=> "read/all"`, to not grant access to protected functions by mistake.

**Comment by Leobouloc** [comment]
> LGTM! 🎉 Note that it could be nice to add or link the list of scopes defined by adlnet LRS specification, maybe in another PR?

Added here: https://github.com/openfun/ralph/blob/bd6fcb615e3a5cba536505f8a6448c234696f369/docs/api.md#scopes !

File: src/ralph/api/auth/user.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (SergioSim) [reply src/ralph/api/auth/user.py:53]: Good point) Ok, thus, we consider the `read/all` scope to be more than just an alias for `{statements,state,profile}/read` scopes. It's clearer to me now) Thanks)
